### PR TITLE
feat: CE-126 Not Authorized Error Page

### PR DESCRIPTION
**Pull request by barrfalk**
# Description

Fixed Authorization on Routes
Updated look and feel of not authorized page

Fixes # (issue)

# How Has This Been Tested?

- [ ] Remove COS Officer role using Common Single Signon, verified "not authorized" page displays
- [ ] Remove TEMPORARY_TEST_ADMIN role using Common Single Signon, verified "not authorized" page displays when accessing the admin/user page.
- [ ] Added roles back, verified access to application

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Further comments

---

Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Backend](https://nr-compliance-enforcement-461-frontend.apps.silver.devops.gov.bc.ca/api/) available
[Frontend](https://nr-compliance-enforcement-461-frontend.apps.silver.devops.gov.bc.ca/) available

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/nr-compliance-enforcement/actions/workflows/merge-main.yml)

---

Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Backend](https://nr-compliance-enforcement-461-frontend.apps.silver.devops.gov.bc.ca/api/) available
[Frontend](https://nr-compliance-enforcement-461-frontend.apps.silver.devops.gov.bc.ca/) available

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/nr-compliance-enforcement/actions/workflows/merge-main.yml)

### CODE REVIEW

File: frontend/src/app/components/routing/protected-routes.tsx [hunk 1]
```diff
@@ -1,15 +1,46 @@
...
+  return requiredRoles.some((role) => userRoles?.includes(role));
+};
+
+export const ProtectedRoutes: FC<ProtectedRoutesProps> = ({ roles }) => {
+  const token = getAuthToken();
+  if (!token) {
+    return <Navigate to="/not-authorized" />;
+  }
+
+  let decodedToken: DecodedToken;
```
Comment (marqueone-ps) [inline frontend/src/app/components/routing/protected-routes.tsx:30]: I like what you've done returning a `<Navigate to="/not-authorized" />;` but you can simplify it a bit more.

```
  const _hasAccess = (requiredRoles: Array<Roles>): boolean => {
    let hasAccess = true;
    const token = localStorage.getItem(AUTH_TOKEN);

    try {
      if (!token) {
        throw Error();
      }

      const { client_roles } = jwtDecode(token) as { client_roles: Array<Roles> };
      if (client_roles.length === 0) {
        hasAccess = false;
      }

      hasAccess = client_roles.some((role) => requiredRoles.includes(role));
    } catch (Error) {
      hasAccess = false;
    } finally {
      return hasAccess;
    }
  };

  if (!_hasAccess(requiredRoles)) {
    return <Navigate to="/not-authorized" />;
  }

  return (
    <Layout>
      <Outlet />
    </Layout>
  );```
  
if you want to keep things as is at the very least you'll want to move your helpers into your component

**Review by marqueone-ps — APPROVED** [review]
Component is working, but could be optimized a bit, but this won't stop from being merged into main

File: frontend/src/app/components/containers/pages/not-authorized.tsx [hunk 2]
```diff
@@ -1,3 +1,29 @@
-export const NotAuthorized = () => {
-  return <h1>You are not authorized to view this page.</h1>;
+import { FC, ReactNode } from "react";
+import { Footer, Header } from "../../containers/layout";
+import { BsPersonFillSlash } from "react-icons/bs";
+
+type props = {
```
[the remark this answers is not in the document]
Comment (barrfalk) [reply frontend/src/app/components/containers/pages/not-authorized.tsx]: Good catch, I've made the change.

File: frontend/src/app/components/routing/protected-routes.tsx [hunk 3]
```diff
@@ -1,15 +1,46 @@
...
 import { Navigate, Outlet } from "react-router-dom";
+import jwt_decode from "jwt-decode";
 import Roles from "../../types/app/roles";
 import Layout from "../containers/layout";
+import { AUTH_TOKEN } from "../../service/user-service";
 
-export const ProtectedRoutes: FC<{ roles: Array<Roles> }> = () => {
-  let auth = { token: true };
-  return auth.token ? (
+interface ProtectedRoutesProps {
```
[the remark this answers is not in the document]
Comment (barrfalk) [reply frontend/src/app/components/routing/protected-routes.tsx]: Changed, thanks!

File: frontend/src/app/components/routing/protected-routes.tsx (hunk 1 above)
Comment (barrfalk) [reply frontend/src/app/components/routing/protected-routes.tsx:26]: I cleaned this up a bit, I _think_ my chages will address your concerns.

**Review by marqueone-ps — APPROVED** [review]

**Pull request merged**